cmd/attest.go

```diff
@@ -0,0 +1,267 @@
+	fmt.Fprint(os.Stderr, "Enter password for private key: ")
+	pw1, err := term.ReadPassword(0)
+	if err != nil {
+		return nil, err
+	}
+	if !confirm {
+		return pw1, nil
+	}
+	fmt.Fprintln(os.Stderr)
+	fmt.Fprint(os.Stderr, "Enter password for private key again: ")
+	confirmpw, err := term.ReadPassword(0)
+	fmt.Fprintln(os.Stderr)
+	if err != nil {
+		return nil, err
+	}
+
+	if string(pw1) != string(confirmpw) {
+		return nil, errors.New("passwords do not match")
+	}
+	return pw1, nil
+}
+
+func passFunc(isPass bool) (b []byte, err error) {
+	pw, ok := os.LookupEnv("COSIGN_PASSWORD")
+	switch {
+	case ok:
+		return []byte(pw), nil
+	case isTerminal():
+		return getPassFromTerm(false)
+	// Handle piped in passwords.
+	default:
+		return io.ReadAll(os.Stdin)
+	}
+}
+
+func validateScheme(userInput string) error {
+	switch {
+	case strings.HasPrefix(userInput, "dir"), strings.HasPrefix(userInput, "file"):
+		return fmt.Errorf("could not support attestation for %s; please try one of these scheme: %v", userInput, validAttestScheme)
+	default:
+		return nil
+	}
+}
+
+func attestExec(ctx context.Context, _ *cobra.Command, args []string) error {
+	// can only be an image for attestation or OCI DIR
+	userInput := args[0]
+	err := validateScheme(userInput)
+	if err != nil {
+		return err
+	}
+
+	ko := sign.KeyOpts{
+		KeyRef:   keyPath,
+		PassFunc: passFunc,
+	}
+
+	sv, err := sign.SignerFromKeyOpts(ctx, "", ko)
+	if err != nil {
+		return err
+	}
+	defer sv.Close()
+
+	return eventLoop(
+		attestationExecWorker(userInput, sv),
+		setupSignals(),
+		eventSubscription,
+		stereoscope.Cleanup,
+		ui.Select(isVerbose(), appConfig.Quiet)...,
+	)
+}
+
+func attestationExecWorker(userInput string, sv *sign.SignerVerifier) <-chan error {
+	errs := make(chan error)
+	go func() {
+		defer close(errs)
+		if len(attestationOutput) > 1 {
+			errs <- fmt.Errorf("can not generate attestation for more than one output")
+			return
+		}
+
+		output := format.ParseOption(attestationOutput[0])
+		if output == format.UnknownFormatOption {
+			errs <- fmt.Errorf("can not use %v as attestation format. Try: %v", output, format.AllOptions)
+			return
+		}
+
+		s, src, err := generateSBOM(userInput, errs)
+		if err != nil {
+			errs <- err
+			return
+		}
+
+		bytes, err := syft.Encode(*s, output)
+		if err != nil {
+			errs <- err
+			return
+		}
+
+		err = generateAttestation(bytes, src, sv, output)
+		if err != nil {
+			errs <- err
+			return
+		}
+	}()
+	return errs
+}
+
+func assertPredicateType(output format.Option) string {
+	switch output {
+	case format.SPDXTagValueOption, format.SPDXJSONOption:
+		return in_toto.PredicateSPDX
+	default:
+		return in_toto.StatementInTotoV01
+	}
+}
+
+func generateAttestation(predicate []byte, src *source.Source, sv *sign.SignerVerifier, output format.Option) error {
+	predicateType := assertPredicateType(output)
+
+	h, _ := v1.NewHash(src.Image.Metadata.ManifestDigest)
+
+	wrapped := dsse.WrapSigner(sv, "application/vnd.in-toto+json")
```

Comment:
nit: could be a good constant (then the name would reflect what this represents)